lib/src/inline_parser.dart

```diff
@@ -449,142 +743,156 @@ class StrikethroughSyntax extends TagSyntax {
-    if (match[3] != null) {
-      // Inline link like [foo](url).
-      var url = match[3];
-      var title = match[4];
-
-      // For whatever reason, Markdown allows angle-bracketed URLs here.
-      if (url.startsWith('<') && url.endsWith('>')) {
-        url = url.substring(1, url.length - 1);
-      }
-
-      return new Link(null, url, title);
-    } else {
-      String id;
-      String _contents() {
-        var offset = pattern.pattern.length - 1;
-        return parser.source.substring(state.startPos + offset, parser.pos);
-      }
-
-      // Reference link like [foo][bar].
-      if (match[1] == null) {
-        // There are no reference brackets ("shortcut reference link"), so infer
-        // the id from the contents.
-        id = _contents();
-      } else if (match[2] == '') {
-        // The id is empty ("[]") so infer it from the contents.
-        id = _contents();
-      } else {
-        id = match[2];
+  // Add a link node to [parser]'s AST.
+  //
+  // If [label] is present, the potential link is treated as a reference link.
+  // Otherwise, it is treated as an inline link.
+  //
+  // Returns whether the image was added successfully.
+  bool _addNode(InlineParser parser, TagState state,
+      {String destination, String title, String label, int endPos}) {
+    Element element;
+    if (label != null) {
+      // Reference link
+      var link = parser.document.refLinks[label];
+      if (link == null) {
+        // This link has no reference definition. But we allow users of the
+        // library to specify a special resolver function ([linkResolver]) that
+        // may choose to handle this. Otherwise, it's just treated as plain
+        // text.
+        if (linkResolver == null) {
+          return false;
+        }
+        var textToResolve = parser.source.substring(state.endPos, parser.pos);
+
+        // See if we have a resolver that will generate a link for us.
+        var node = linkResolver(textToResolve);
+        if (node == null) {
+          return false;
+        } else {
+          parser.addNode(node);
+          parser.start = parser.pos = endPos;
+          return true;
+        }
       }
+      destination = link.url;
+      title = link.title;
+    }
 
-      // References are case-insensitive.
-      id = id.toLowerCase();
-      return parser.document.refLinks[id];
+    element = new Element('a', state.children);
+    element.attributes['href'] = escapeAttribute(
+        destination.replaceAll(r'\(', '(').replaceAll(r'\)', ')'));
+    if (title != null && title.isNotEmpty) {
+      element.attributes['title'] = escapeAttribute(title);
     }
-  }
+    parser.addNode(element);
+    parser.start = parser.pos = endPos;
 
-  bool onMatchEnd(InlineParser parser, Match match, TagState state) {
-    var node = createNode(parser, match, state);
-    if (node == null) return false;
+    // Mark all of the `[` tag states as inactive.
+    parser.deactivatePendingLinks();
```

Comment:
What problem does this solve? It would be good to document.

Reply:
Done.